openff/toolkit/utils/exceptions.py

```diff
@@ -1,48 +1,45 @@
-__all__ = (
-    "MessageException",
-    "IncompatibleUnitError",
-    "MissingDependencyError",
-    "MissingPackageError",
-    "ToolkitUnavailableException",
-    "LicenseError",
-    "InvalidToolkitError",
-    "InvalidToolkitRegistryError",
-    "UndefinedStereochemistryError",
-    "GAFFAtomTypeWarning",
-    "ChargeMethodUnavailableError",
-    "IncorrectNumConformersError",
-    "IncorrectNumConformersWarning",
-    "ChargeCalculationError",
-    "InvalidIUPACNameError",
-    "AntechamberNotFoundError",
-    "ParseError",
-)
-
-# =============================================================================================
-# COMMON EXCEPTION TYPES
-# =============================================================================================
-
-
-class MessageException(Exception):
-    """A base class for exceptions that print out a string given in their constructor"""
+import warnings
+
+deprecated_names = ["MessageException", "ParseError"]
+
+
+def __getattr__(name):
+    if name in deprecated_names:
+        warnings.filterwarnings("default", category=DeprecationWarning)
+        warning_msg = f"{name} is DEPRECATED and will be removed in a future release of the OpenFF Toolkit."
+        if name == "MessageException":
+            warning_msg += (
+                " All custom exceptions now inherit from OpenFFToolkitException, "
+                "which should be used as a replacement for MessageException. Import it via "
+                "`from openff.toolkit.utils.exceptions import OpenFFToolkitException`."
+            )
+        warnings.warn(warning_msg, DeprecationWarning)
+        return globals()[f"_Deprecated{name}"]
+    raise AttributeError(f"module {__name__} has no attribute {name}")
+
+
+class _DeprecatedMessageException(Exception):
+    """DEPRECATED: A base class for exceptions that print out a string given in their constructor"""
+
+
+class OpenFFToolkitException(Exception):
+    """Base exception for custom exceptions raised by the OpenFF Toolkit"""
 
     def __init__(self, msg):
-        super().__init__(self, msg)
+        super().__init__(msg)
         self.msg = msg
 
     def __str__(self):
         return self.msg
```

Comment:
I'm curious what additional functionality this adds to `Exception`? Is it having the `.msg` attribute?

```python
>>> exc = ValueError("my message")
>>> str(exc)
'my message'
>>> exc.args
('my message',)
```

Reply:
Pretty much, yeah. I could go either way on this; I recall seeing/using some code that relies on `.msg` existing, but I couldn't find it with a quick. I might prefer the aesthetic of clearly having a message attribute over hoping that calling `str()` on it. I think that should always work, but I vaguely recall being bitten by cases in which that isn't true.

Reply:
It'll give you functional differences if `msg` ever happens to not be a string, I'm not sure if that'd affect scripts and things. `OFFTKExc.msg` would just naively return 3, I assume.

```python
>>> exc = ValueError(3)
>>> str(exc)
'3'
>>> exc.args
(3,)
```

Normal exceptions also take in multiple arguments.

```python
>>> exc2 = ValueError("This is probably good for complex errors", "but I've never seen anyone use it", 3)
>>> str(exc2)
'(\'This is probably good for complex errors\', "but I\'ve never seen anyone use it", 3)'
>>> exc2.args
('This is probably good for complex errors', "but I've never seen anyone use it", 3)
>>> raise exc2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: ('This is probably good for complex errors', "but I've never seen anyone use it", 3)
```

Reply:
Hmm, I don't know why I dropped the `self` compared to the same code in `MessageException`; it looks out of place, but I probably did it at some point to get an ephemeral block of code to do what I wanted. The inside of `OpenFFToolkitException` is otherwise copied from `MessageException`, I didn't put care into thinking if that was wise.

What I'm gathering from this code is that ... overriding `__init__` and `__str__` is unnecessary? Whether writing tests or being a user of the toolkit, I'd be ok with patterns like

```python3
try:
    do_toolkit_thing(...)
except SomeError as e:
    if "foo" in str(e):
        ...
    elif "bar" in str(e):
        ...
```

which is functionally the same as `if "foo" in e.msg` but doesn't override anything while subclassing out of the stdlib